lxd/images.go

```diff
@@ -1972,34 +1977,92 @@ func imageExport(d *Daemon, r *http.Request) response.Response {
 	return response.FileResponse(r, files, nil, false)
 }
 
-func imageSecret(d *Daemon, r *http.Request) response.Response {
+func imageExportPost(d *Daemon, r *http.Request) response.Response {
 	project := projectParam(r)
 	fingerprint := mux.Vars(r)["fingerprint"]
-	_, imgInfo, err := d.cluster.ImageGet(project, fingerprint, false, false)
+
+	req := api.ImageExportPost{}
+
+	err := json.NewDecoder(r.Body).Decode(&req)
 	if err != nil {
-		return response.SmartError(err)
+		response.SmartError(err)
 	}
 
-	secret, err := shared.RandomCryptoString()
+	// Connect to the target and push the image
+	args := &lxd.ConnectionArgs{
+		TLSServerCert: req.Certificate,
+		UserAgent:     version.UserAgent,
+		Proxy:         d.proxy,
+		CachePath:     d.os.CacheDir,
+		CacheExpiry:   time.Hour,
+	}
 
+	// Setup LXD client
+	remote, err := lxd.ConnectLXD(req.Target, args)
 	if err != nil {
-		return response.InternalError(err)
+		return response.SmartError(err)
 	}
 
-	meta := shared.Jmap{}
-	meta["secret"] = secret
+	run := func(op *operations.Operation) error {
+		createArgs := &lxd.ImageCreateArgs{}
+		imageMetaPath := shared.VarPath("images", fingerprint)
+		imageRootfsPath := shared.VarPath("images", fingerprint+".rootfs")
 
-	resources := map[string][]string{}
-	resources["images"] = []string{imgInfo.Fingerprint}
+		metaFile, err := os.Open(imageMetaPath)
+		if err != nil {
+			return err
+		}
+		defer metaFile.Close()
 
-	op, err := operations.OperationCreate(d.State(), project, operations.OperationClassToken, db.OperationImageToken, resources, meta, nil, nil, nil)
+		createArgs.MetaFile = metaFile
+		createArgs.MetaName = filepath.Base(imageMetaPath)
+
+		if shared.PathExists(imageRootfsPath) {
+			rootfsFile, err := os.Open(imageRootfsPath)
+			if err != nil {
+				return err
+			}
+			defer rootfsFile.Close()
+
+			createArgs.RootfsFile = rootfsFile
+			createArgs.RootfsName = filepath.Base(imageRootfsPath)
+		}
+
+		image := api.ImagesPost{}
+		image.Filename = createArgs.MetaName
+		image.Source = &api.ImagesPostSource{
+			Fingerprint: fingerprint,
+			Secret:      req.Secret,
+		}
+
+		_, err = remote.CreateImage(image, createArgs)
```

Comment:
It would probably make sense to either wait for `CreateImage` to be done (`op.Wait`) or make the operation cancelable. However, the source is only trusted by the target once (via the token which gets invalidated immediately), and will therefore fail when trying to do anything further, e.g. retrieve any operations.

Since the client is trusted by both the source and the target, can't the client just add the source as a trusted "client" (or even server) to the target, either permanently or for the duration of the entire operation? We could have the client set the `type` field to "server" to make the target aware, and perhaps only allow the source to perform certain tasks.

@stgraber what do you think?

Reply:
Or we could just keep the token valid until everything is done.